app/src/main/java/com/hyperether/getgoing/activity/ShowDataActivity.java

```diff
@@ -61,6 +73,114 @@ protected void onListItemClick(ListView l, View v, int position, long id) {
+        }
+        charDates.addAll(charDatesSet);
+        return charDates;
+    }
+
+    private ArrayList<Map<String, Float>> collectData() {
+        ArrayList<Map<String, Float>> collectedData = new ArrayList<>();
+        Map<String, Float> mapY1 = new LinkedHashMap<>();
+        Map<String, Float> mapY2 = new LinkedHashMap<>();
+        Map<String, Float> mapY3 = new LinkedHashMap<>();
+
+        for (DbRoute item : routes) {
+            if (mapY1.containsKey(item.getDate().substring(0, 10)) && item.getActivity_id() == 1) {
+                float q = mapY1.get(item.getDate().substring(0, 10)) + (float) item.getEnergy();
+                mapY1.put(item.getDate().substring(0, 10), q);
+            } else if (item.getActivity_id() == 1) {
+                mapY1.put(item.getDate().substring(0, 10), (float) item.getEnergy());
+            } else if (mapY2.containsKey(item.getDate().substring(0, 10)) && item.getActivity_id() == 2) {
+                float q = mapY2.get(item.getDate().substring(0, 10)) + (float) item.getEnergy();
+                mapY2.put(item.getDate().substring(0, 10), q);
+            } else if (item.getActivity_id() == 2) {
+                mapY2.put(item.getDate().substring(0, 10), (float) item.getEnergy());
+            } else if (mapY3.containsKey(item.getDate().substring(0, 10)) && item.getActivity_id() == 3) {
+                float q = mapY3.get(item.getDate().substring(0, 10)) + (float) item.getEnergy();
+                mapY3.put(item.getDate().substring(0, 10), q);
+            } else if (item.getActivity_id() == 3) {
+                mapY3.put(item.getDate().substring(0, 10), (float) item.getEnergy());
+            }
+        }
+        collectedData.add(mapY1);
+        collectedData.add(mapY2);
+        collectedData.add(mapY3);
+        return collectedData;
+    }
+
+    private ArrayList<IBarDataSet> prepareData() {
+        ArrayList<IBarDataSet> dataSets = new ArrayList<>();
+        ArrayList<BarEntry> yVal1 = new ArrayList<>();
+        ArrayList<BarEntry> yVal2 = new ArrayList<>();
+        ArrayList<BarEntry> yVal3 = new ArrayList<>();
+        ArrayList<String> x = xValuesChart();
+        Map<String, Float> y1 = collectData().get(0);
+        Map<String, Float> y2 = collectData().get(1);
+        Map<String, Float> y3 = collectData().get(2);
+        String xVal;
+        float emptyValue = 0;
+
+        for (int i = 0; i < x.size(); i++) {
+            xVal = x.get(i);
+            if (y1.containsKey(x.get(i))) {
+                yVal1.add(new BarEntry(y1.get(xVal), i));
+            } else {
+                yVal1.add(new BarEntry(emptyValue, i));
+            }
+        }
+        for (int i = 0; i < x.size(); i++) {
+            xVal = x.get(i);
+            if (y2.containsKey(x.get(i))) {
+                yVal2.add(new BarEntry(y2.get(xVal), i));
+            } else {
+                yVal2.add(new BarEntry(emptyValue, i));
+            }
+        }
```

Comment:
@NikolaHadzicNs is there a reason for having 3 loops for x array: for (int i = 0; i < x.size(); i++)  ? Why not only one?

Reply:
I had a problem with group of bars when i was using one loop probably because empty values. Now it is good, i set one loop instead of 3. Thanks for showing me directions.